site/docs/types/type_system.md

```diff
@@ -0,0 +1,36 @@
+# Type System
+
+Substrait tries to cover the most common types used in data manipulation. To minimize type explosion, the project currently follows the guideline that a logical type should probably only be included in the specification if it is included in at least two of the following open source Projects: Apache Arrow, Apache Iceberg, Apache Spark and Trino. Types beyond this common core may be represented using [simple extensions](../extensions/index.md#simple-extensions).
+
+## Logical vs. Physical Types
```

Comment:
As above, I think you're using a specific version of these terms which doesn't map well to relational algebra land. As mentioned elsewhere, all types in substrait carry type variation. It was an early concrete change in Substrait to avoid a separation between logical and physical type definitions. This section is useful but I'd title this section "Type Interpretation Guidelines" or something similar.

Reply:
Terminology aside, I'm getting really confused now. If variations don't relate to different ~~physical~~ storage methods of a ~~logical~~ type, then what are they? Just some kind of inheritance-like thing or something? Or is the word "physical" merely overloaded differently in my mind than it is in yours due to relational algebra stuff?

Also, does

> all types in substrait carry type variation

imply that variations aren't optional...? Because then every single plan I've ever seen is invalid (so needless to say I think that would be a bad idea), and the validator is wrong about a lot of things.

Reply:
Yeah, I'd say the docs aren't super clear on this and it is good that you're trying to clean things up. Let's start with the simple case:

Many systems have a single canonical representation of each type of data. In those systems, people can use the default type variation of 0 meaning "system default".  

In some systems, a system could have multiple variations of representation for the same data. For example, I may have a normal i64 and a rle-encoded i64. In this case, the system would decide which of the type variations each of those things are mapped to. E.g. `i64[0] => normal` and `i64[1] => rle-encoded`.

> that variations aren't optional...?

i64 is the same as i64[0]. So they are optional to specify, they just default 0. The property always exists and has a value. This is the "default" variation for the simple case where systems only have one variation. Protobuf hides default values by default and thus 0 is implied rather than being concrete when sending on the wire. It's absence is the same as it's presence with value 0. In the simple case, the default equals "there is no variation". One thing that is missing is that we should formally specify the properties of the 0 type variation. Specifically it has `function behavior == inherits`. This was implied but isn't written anywhere.

Put another way, types always have a variation property. If not expressly stated, they default to 0, which means "system default".

Reply:
Okay, so (the integer vs name thing from the other thread aside) I *think* this just boils down to us misunderstanding each other, but both understanding the actual spec the same way. Phew :)